crates/edr_napi/src/trace/vm_trace_decoder.rs

```diff
@@ -150,22 +134,17 @@ impl VmTraceDecoder {
 
     #[napi]
     pub fn get_contract_and_function_names_for_call(
-        &self,
+        &mut self,
```

Comment:
To be honest I didn't dig deep enough in this particular case (`&mut self` in associated functions) but I'll make an educated guess that this may be prone to the reentrancy issues that we discovered with `ClassInstance`/`Reference`.

To recap, the class instance data is owned by V8/napi and so the glue code probably couples that with the assumption about being single-threaded and assumes that it's safe to give out exclusive access in Rust to data that is only being managed by V8, since it wouldn't be possible to manifest that reference safely in Rust by the user. However, it still might be possible to call into this using regular JS object API, which would manifest a second `&mut` reference, which is UB.

Since we do not do that ourselves, i.e. calling internally any of the `napi-rs` exported logic ourselves inside this Rust code, we should be safe :pray:

Reply:
Thanks for the explanation